### PR TITLE
test: Introduce snapshot testing for test-utils selectors

### PR DESCRIPTION
**Pull request by pan-kot**
### Description

A new test for test-utils selectors extracts all referenced selectors, groups them by component, and captures a snapshot. It makes it visible when selectors used in test-utils are updated which might cause a mismatch between components and test infrastructure in case different versions of the package are used.

The selectors copy the trim-hash approach from https://github.com/cloudscape-design/test-utils/blob/main/packages/core/src/utils.ts#L51. We can consider adding the utility to extract selectors to the test-utils package to be consumed in the test.

See: https://sim.amazon.com/issues/AWSUI-22332

### How has this been tested?

<details>
   <summary>Review checklist</summary>

_The following items are to be evaluated by the author(s) and the reviewer(s)._

#### Correctness

- _Changes include appropriate documentation updates._
- _Changes are backward-compatible if not indicated, see [`CONTRIBUTING.md`](https://github.com/cloudscape-design/components/blob/main/CONTRIBUTING.md#public-apis)._
- _Changes do not include unsupported browser features, see [`CONTRIBUTING.md`](https://github.com/cloudscape-design/components/blob/main/CONTRIBUTING.md#browsers-support)._
- _Changes were manually tested for accessibility, see [accessibility guidelines](https://cloudscape.design/foundation/core-principles/accessibility/)._

#### Security

- _If the code handles URLs: all URLs are validated through [the `checkSafeUrl` function](https://github.com/cloudscape-design/components/blob/main/src/internal/utils/check-safe-url.ts)._

#### Testing

- _Changes are covered with new/existing unit tests?_
- _Changes are covered with new/existing integration tests?_
</details>

---

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: package-lock.json [hunk 1]
```diff
@@ -14207,9 +14207,9 @@
       }
```
Comment (abdhalees) [inline package-lock.json:14207]: are these changes necessary?

Comment (pan-kot) [reply package-lock.json:14207]: No, those are just resulted from running npm install - can remove those from the pr

File: src/test-utils/dom/collection-preferences/sticky-columns-preference.ts [hunk 2]
```diff
@@ -10,12 +10,12 @@ export default class StickyColumnsPreferenceWrapper extends ComponentWrapper {
   static lastRootSelector = styles['sticky-columns-last'];
 
   findTitle(): ElementWrapper {
-    return this.findComponent(`.${styles[`sticky-columns-form-field`]}`, FormFieldWrapper)!.findLabel()!;
+    return this.findComponent(`.${styles['sticky-columns-form-field']}`, FormFieldWrapper)!.findLabel()!;
```
Comment (just-boris) [inline src/test-utils/dom/collection-preferences/sticky-columns-preference.ts:13]: Why is this needed? What happens with the extractor when a template string used?

File: src/__tests__/test-utils-selectors.test.tsx [hunk 3]
```diff
@@ -0,0 +1,141 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+import fs from 'fs';
+import Path from 'path';
+import glob from 'glob';
+import { transformSync, types, PluginObj, NodePath } from '@babel/core';
```
Comment (just-boris) [inline src/__tests__/test-utils-selectors.test.tsx:7]: This module is not declared in our dependencies. You should `npm install --save-dev @babel/core` to not break our builds.

File: src/__tests__/test-utils-selectors.test.tsx [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+import { flatten, zip } from 'lodash';
+
+// The test extracts generated selectors from the compiled output and matches those against the snapshot.
+test('test-utils selectors', () => {
+  // Find referenced selector files and properties.
+  const selectorsFilePathToUsedProperties = new Map<string, Set<string>>();
+  for (const file of glob.sync('lib/components/test-utils/selectors/**/*.js')) {
```
Comment (just-boris) [inline src/__tests__/test-utils-selectors.test.tsx:14]: Are you sure it is a good idea to exctract this info from the compiled code? Would it be easier to extract it from typescript sources? (`import` syntax is much easier to statically analyse, I can see how the code below simplifies)

File: src/__tests__/test-utils-selectors.test.tsx [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+    const selectorVars = new Map<string, string>();
+    return {
+      visitor: {
+        // Find require statements that import selectors.
+        CallExpression(path: NodePath<types.CallExpression>) {
+          if (path.node.callee.type === 'Identifier' && path.node.callee.name === 'require') {
+            const argument = path.node.arguments[0];
+            if (argument.type === 'StringLiteral' && argument.value.endsWith('selectors.js')) {
+              if (path.parent.type === 'VariableDeclarator' && path.parent.id.type === 'Identifier') {
+                selectorVars.set(path.parent.id.name, Path.resolve(file, '..', argument.value));
```
Comment (just-boris) [inline src/__tests__/test-utils-selectors.test.tsx:44]: Why is there an extra `..`? At least it needs an explaining comment

File: src/__tests__/test-utils-selectors.test.tsx [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+              }
+            }
+          }
+        },
+        // Find selector references and extract used property names.
+        MemberExpression(path: NodePath<types.MemberExpression>) {
+          function parseNested(expression: types.MemberExpression, propertyName: string) {
+            if (expression.object.type === 'Identifier' && selectorVars.has(expression.object.name)) {
+              const filePath = selectorVars.get(expression.object.name)!;
```
Comment (just-boris) [inline src/__tests__/test-utils-selectors.test.tsx:53]: This fails if it ever hits this code


```js
import styles from '...../selectors.js'

findSomething() {
   const styles = getSomethingElse();
   return this.find(styles[....]); // this will be incorrectly classified as coming from the import
}
```

Babel has functionality to track variables and where they are coming from: https://github.com/jamiebuilds/babel-handbook/blob/master/translations/en/plugin-handbook.md#toc-scopes

Use scopes to track this properly

File: src/__tests__/test-utils-selectors.test.tsx [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+    if (!element) {
+      continue;
+    } else if (element.type === 'TemplateElement') {
+      literal += element.value.raw;
+    } else if (element.type === 'Identifier') {
+      literal += '*';
+    } else if (element.type === 'NumericLiteral') {
+      literal += element.value;
+    } else {
+      throw new Error('Unhandled template literal structure.');
```
Comment (just-boris) [inline src/__tests__/test-utils-selectors.test.tsx:124]: Can we also add file name and line number to the error message to make it easier for other devs to investigate the issues?

Same applies to all errors here. Also, another reason to switch to the source parsing, to make it easier to track down issues.

File: src/__tests__/test-utils-selectors.test.tsx [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+    }
+  }
+  return literal;
+}
+
+// Match property against the used properties supporting the wildcards ("*").
```
Comment (just-boris) [inline src/__tests__/test-utils-selectors.test.tsx:130]: In which condition is there a wildcard?

File: src/test-utils/dom/collection-preferences/sticky-columns-preference.ts (hunk 2 above)
Comment (pan-kot) [reply src/test-utils/dom/collection-preferences/sticky-columns-preference.ts:13]: This change is not required. The templates strings like these ones are handled just fine. If a template string includes an identifier it is then replaced with a wildcard, e.g. `flash-type-*` to match all properties starting from `flash-type-`.

File: src/__tests__/test-utils-selectors.test.tsx (hunk 4 above)
Comment (pan-kot) [reply src/__tests__/test-utils-selectors.test.tsx:14]: Thanks, I will try this out.

File: src/__tests__/test-utils-selectors.test.tsx (hunk 5 above)
Comment (pan-kot) [reply src/__tests__/test-utils-selectors.test.tsx:44]: Thanks, I will add a comment

File: src/__tests__/test-utils-selectors.test.tsx (hunk 8 above)
Comment (pan-kot) [reply src/__tests__/test-utils-selectors.test.tsx:130]: Explained here: https://github.com/cloudscape-design/components/pull/1863#discussion_r1444768436

File: src/__tests__/test-utils-selectors.test.tsx (hunk 7 above)
Comment (pan-kot) [reply src/__tests__/test-utils-selectors.test.tsx:124]: This should be visible in the stack trace. I also plan to move the extractor code to the test-utils package eventually so it can be reused across packages.

Comment (just-boris) [reply src/__tests__/test-utils-selectors.test.tsx:124]: > This should be visible in the stack trace

Can you show an example?

File: src/test-utils/dom/collection-preferences/sticky-columns-preference.ts (hunk 2 above)
Comment (just-boris) [reply src/test-utils/dom/collection-preferences/sticky-columns-preference.ts:13]: Let's just throw an error saying that dynamic property access is not allowed and simplify everything

Comment (pan-kot) [reply src/test-utils/dom/collection-preferences/sticky-columns-preference.ts:13]: I think it works well so far. An alternative could be to update the source code of the test-utils, for example, instead of `styles[`flash-type-${type}`]` we can do:

```
function getFlashStyleByType(type) {
  switch (type) {
    case 'error':
      return styles.error;
    case 'in-progress':
      return styles['in-progress'];
    // ...
  }
}
```

But I would rather keep the current solution with the extractor.

File: src/__tests__/test-utils-selectors.test.tsx (hunk 7 above)
Comment (pan-kot) [reply src/__tests__/test-utils-selectors.test.tsx:124]: Ah, no. Sorry, I misunderstood the suggestion. I will add file path and line number to every thrown exception.

File: src/test-utils/dom/collection-preferences/sticky-columns-preference.ts (hunk 2 above)
Comment (just-boris) [reply src/test-utils/dom/collection-preferences/sticky-columns-preference.ts:13]: Do we have any need for this case today? If not, why do you need to think about it?

Comment (pan-kot) [reply src/test-utils/dom/collection-preferences/sticky-columns-preference.ts:13]: Yes, there are 22 selectors matched this way:
```
"awsui_content-display-description_tc96w",
"awsui_content-display-option-content_tc96w",
"awsui_content-display-option-list_tc96w",
"awsui_content-display-option_tc96w",
"awsui_content-display-title_tc96w",
"awsui_visible-content-group-label_tc96w",
"awsui_visible-content-group_tc96w",
"awsui_visible-content-groups_tc96w",
"awsui_visible-content-option-label_tc96w",
"awsui_visible-content-option_tc96w",
"awsui_visible-content-title_tc96w",
"awsui_visible-content-toggle_tc96w",
"awsui_flash-type-error_1q84n",
"awsui_flash-type-in-progress_1q84n",
"awsui_flash-type-info_1q84n",
"awsui_flash-type-success_1q84n",
"awsui_flash-type-warning_1q84n",
"awsui_result-container-error_11huc",
"awsui_result-container-success_11huc",
"awsui_navigation-link-active_1xupv",
"awsui_navigation-link-disabled_1xupv",
"awsui_navigation-link-label_1xupv",
```

File: src/__tests__/test-utils-selectors.test.tsx [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+function matchProperties(usedProperties: string[], property: string) {
+  for (const testProperty of usedProperties) {
+    if (testProperty === property) {
+      return true;
+    }
+    if (testProperty.includes('*') && new RegExp(testProperty.replace(/\*/g, '.*')).test(property)) {
```
Comment (just-boris) [inline src/__tests__/test-utils-selectors.test.tsx:136]: Dynamically constructing regexp is unsafe, because there could be other special characters which can be treated as control statements. 

Learn how to write safe regexps and update

File: src/__tests__/test-utils-selectors.test.tsx (hunk 6 above)
Comment (pan-kot) [reply src/__tests__/test-utils-selectors.test.tsx:53]: This actually fails in the described scenario when running the script against uncompiled sources (in the compiled ones the var names are generated). However, in the worst case scenario the output would contain a few extra selectors which is fine. The case is extremely unlikely though: not only the variable name has to be duplicated, the accessed property must exist. I propose to not introduce scopes tracking unless there is a simple way of adding it.

File: src/__tests__/test-utils-selectors.test.tsx (hunk 9 above)
Comment (pan-kot) [reply src/__tests__/test-utils-selectors.test.tsx:139]: How much do we need it for this case? Yes, there can be in theory some weird template string so that the resulting expression will match more or less properties in the worst scenario.

Comment (just-boris) [reply src/__tests__/test-utils-selectors.test.tsx:139]: If you want to have a weaknesses in your solution, I cannot do anything with that

Comment (just-boris) [reply src/__tests__/test-utils-selectors.test.tsx:139]: But at least you could add `^` and `$` to ensure it matches the full strings only

File: src/__tests__/test-utils-selectors.test.tsx [hunk 10]
```diff
@@ -0,0 +1,144 @@
...
+  const splitSelector = selector.replace('.', '').split('_');
+  if (splitSelector.length >= 5) {
+    splitSelector.splice(splitSelector.length - 2, splitSelector.length);
+    return splitSelector.join('_');
+  }
+  return selector;
+}
+
+function resolveSelectorsPath(importPath: string) {
+  return 'lib/components/' + importPath.replace(/\.\.\//g, '');
```
Comment (just-boris) [inline src/__tests__/test-utils-selectors.test.tsx:125]: Still unclear what is happening here. Why did you not use `path.resolve(from, to)` to follow relative paths?

File: src/__tests__/test-utils-selectors.test.tsx (hunk 6 above)
Comment (just-boris) [reply src/__tests__/test-utils-selectors.test.tsx]: This is what you needed to do


```js
export default function (babel) {
  const { types: t } = babel;

  return {
    name: "ast-transform", // not required
    visitor: {
      MemberExpression(path) {
        const binding = path.scope.getBinding(path.node.object.name);
        if (binding.kind !== 'module' || !binding.path.parent.source.value.endsWith('selectors.js')) {
          return;
        }
        // this is an expected import
      }
    }
  };
}
```

Live demo: https://astexplorer.net/#/gist/9fcb7096d184ca40423225db5f41baeb/1f4d1737b0dfd6fb7fa025ec9da416f3dd1a1f93

Do you still think it is difficult?

Comment (pan-kot) [reply src/__tests__/test-utils-selectors.test.tsx]: Many thanks for the reference, I was not familiar with getBinding API. I will add the code

File: src/__tests__/test-utils-selectors.test.tsx (hunk 10 above)
Comment (pan-kot) [reply src/__tests__/test-utils-selectors.test.tsx:125]: The import path is taken from `src/test-utils/selectors/{component}`. The selectors path should resolve to `lib/components/{component}/styles.selectors.js`.

If using `Path.resolve` this can be achieved with e.g.
```
Path.resolve(testUtilFilePath, importPath).replace('/src/', '/lib/')
```

I like the current approach better.

Though, this needs to be parametrised somewhat when moving the extractor part to the test-utils package so that it can be reused.

Comment (pan-kot) [reply src/__tests__/test-utils-selectors.test.tsx:125]: The import path is taken from `src/test-utils/selectors/{component}`. The selectors path should resolve to `lib/components/{component}/styles.selectors.js`.

If using `Path.resolve` this can be achieved with e.g.
```
Path.resolve(testUtilFilePath, importPath).replace('/src/', '/lib/')
```

I like the current approach better.

Though, this needs to be parametrised somewhat when moving the extractor part to the test-utils package so that it can be reused.

Comment (pan-kot) [reply src/__tests__/test-utils-selectors.test.tsx:125]: The solution does support nested paths like `../../../app-layout/test-classes/styles.selectors.js` but I will give your solution a try

File: src/__tests__/test-utils-selectors.test.tsx [hunk 11]
```diff
@@ -0,0 +1,162 @@
...
+          }
+        },
+      },
+    } as PluginObj;
+  }
+  const source = fs.readFileSync(file, 'utf-8');
+  transformSync(source, { babelrc: false, configFile: false, plugins: [...defaultPlugins, extractor] })?.code;
+}
+
+function extractComponentSelectors(file: string, usedProperties: string[], onExtract: (selector: string) => void) {
```
[the remark this answers is not in the document]
Comment (pan-kot) [reply src/__tests__/test-utils-selectors.test.tsx:92]: Updated

File: src/__tests__/test-utils-selectors.test.tsx [hunk 12]
```diff
@@ -0,0 +1,162 @@
...
+        ImportDeclaration(path: NodePath<types.ImportDeclaration>) {
+          if (path.node.source.value.endsWith('selectors.js') && path.node.specifiers.length === 1) {
+            const specifier = path.node.specifiers[0];
+            if (specifier.type === 'ImportDefaultSpecifier') {
+              selectorToFilePath.set(specifier.local.name, resolveSelectorsPath(path.node.source.value));
+            } else {
+              throw new Error(`Unhandled styles import type at ${file}:${path.node.loc?.start.line}.`);
+            }
+          }
+        },
```
[the remark this answers is not in the document]
Comment (pan-kot) [reply src/__tests__/test-utils-selectors.test.tsx]: Updated

File: src/__tests__/test-utils-selectors.test.tsx (hunk 10 above)
Comment (pan-kot) [reply src/__tests__/test-utils-selectors.test.tsx]: Updated
